src/main/java/com/example/petbuddybackend/service/chat/ChatService.java

```diff
@@ -61,14 +64,19 @@ public Page<ChatMessageDTO> getChatMessagesByParticipantEmail(Long chatId,
     }
 
     @Transactional(readOnly = true)
-    public Page<ChatRoomDTO> getChatRoomsByParticipantEmailSortedByLastMessage(String principalEmail,
-                                                                               Role role,
-                                                                               Pageable pageable,
-                                                                               ZoneId timeZone) {
-        Page<ChatMessage> principalChatMessages = role == Role.CLIENT ?
-                chatMessageRepository.findLastMessagesOfChatRoomsOfClient(principalEmail, pageable) :
-                chatMessageRepository.findLastMessagesOfChatRoomsOfCaretaker(principalEmail, pageable);
+    public Page<ChatRoomDTO> getChatRoomsByCriteriaSortedByLastMessage(String principalEmail,
+                                                                       Role role,
+                                                                       ChatRoomSearchCriteria searchCriteria,
+                                                                       Pageable pageable,
+                                                                       ZoneId timeZone) {
+
+        Specification<ChatMessage> chatMessageSpec = ChatSpecificationUtils.filtersToSpecificationSorted(
+                searchCriteria,
+                principalEmail,
+                role
+        );
 
+        Page<ChatMessage> principalChatMessages = chatMessageRepository.findAll(chatMessageSpec, pageable);
         return principalChatMessages.map(message -> mapToChatRoomDTO(principalEmail, message, timeZone));
```

Comment:
```suggestion
        return principalChatMessages.map(message -> mapToChatRoomDTO(principalEmail, message, timeZone));
```
return chatMessageRepository.findAll(chatMessageSpec, pageable)
                .map(message -> mapToChatRoomDTO(principalEmail, message, timeZone));

Reply:
Dodatkowo troche to takie na okolo. Moze od razu findAll zrobic w chatRepository i pobrac czaty, bo w koncu chcemy pobrac czaty i w mapToChatRoomDTO i tak je pobieramy. Do messages w ChatRoom mozna dac @OrderBy, wiec mielibysmy na 1. miejscu ostatnia wiadomosc.
Minus jest taki, ze wtedy trzeba zaciagnac wszystkie wiadomosci czatu, zeby zebrac jedną.

Ale drugą opcją mogloby byc to, ze najpierw wywolamy findAll do wyciagniecia czatow, a potem w metodzie mapToChatRoomDTO wywolalibysmy chatMessageRepository i tam przez query wyciagneli ostatnia wiadomosc danego czatu. Byloby to juz wydajniejsze niz 1. metoda i wydaje mi sie, ze czytelniejsze  niz to co mamy obecnie.

Chociaz to co mamy obecnie jest pewnie najbardziej wydajne.

Reply:
>  Minus jest taki, ze wtedy trzeba zaciagnac wszystkie wiadomosci czatu, zeby zebrac jedną.

No właśnie dlatego bym może nie używał pierwszej metody

> Ale drugą opcją mogloby byc to, ze najpierw wywolamy findAll do wyciagniecia czatow, a potem w metodzie mapToChatRoomDTO wywolalibysmy chatMessageRepository i tam przez query wyciagneli ostatnia wiadomosc danego czatu.

Ale trzeba sortować po ostatniej wiadomości na poziomie bazy. W przeciwnym wypadku musiałbym wszyskie czaty wyciągać

Reply:
Wiem że z specyfikacją troche się skomplikowało. Teraz myślę że mogłem po prostu do `@Query` dopisać pare linijek i byłoby to samo, ale jak już napisałem tak to chyba zostawię. Plus na pewno tego taki że jest to łatwiej rozszerzalne w przeciwieństwie do `@Query`

Reply:
> > Minus jest taki, ze wtedy trzeba zaciagnac wszystkie wiadomosci czatu, zeby zebrac jedną.
> 
> No właśnie dlatego bym może nie używał pierwszej metody
> 
> > Ale drugą opcją mogloby byc to, ze najpierw wywolamy findAll do wyciagniecia czatow, a potem w metodzie mapToChatRoomDTO wywolalibysmy chatMessageRepository i tam przez query wyciagneli ostatnia wiadomosc danego czatu.
> 
> Ale trzeba sortować po ostatniej wiadomości na poziomie bazy. W przeciwnym wypadku musiałbym wszyskie czaty wyciągać

W specification tez sortujesz te wiadomosci

Reply:
Ale na poziomie bazy danych

Reply:
> Ale drugą opcją mogloby byc to, ze najpierw wywolamy findAll do wyciagniecia czatow, a potem w metodzie mapToChatRoomDTO wywolalibysmy chatMessageRepository i tam przez query wyciagneli ostatnia wiadomosc danego czatu. Byloby to juz wydajniejsze niz 1. metoda i wydaje mi sie, ze czytelniejsze niz to co mamy obecnie.

Bo jak dobrze rozumiem, to chcesz wyciągnąć wszyskie czaty i wtedy do nich dociągnąć wiadomość i wtedy posortować ale na backendzie?

Reply:
nie nie. Wyciagnac wszytkie czaty i potem jak mamy mapToChatRoomDTO to tam wywolywac query na chatRoomMessageRepository, ktoreby zwriclo ostatnia wiadomosc z czatu, wiec tez bysmy po bazie sortowali

Reply:
To w sumie podobnie jak wcześniej, tylko by trzeba było to wyszukiwanie po danych wcisnąć

Reply:
Ale też wszyskich bym nie wyciągał czatów tylko te co Pageable wskazuje

Reply:
To te co pageable, chodzi tylko o to żeby wyciągnąć czaty i potem mapujac wyciągnąć ostatnia wiadomość czatu przez query

Reply:
```java
    @Query("""
        SELECT cm
        FROM ChatMessage cm
        WHERE cm.chatRoom.caretaker.email = :caretakerUsername
            AND cm.createdAt = (
                SELECT MAX(cm2.createdAt)
                FROM ChatMessage cm2
                WHERE cm2.chatRoom = cm.chatRoom
            )
            AND (
                cm.chatRoom.caretaker.email LIKE CONCAT('%', :chatterData, '%')
                OR cm.chatRoom.caretaker.accountData.name LIKE CONCAT('%', :chatterData, '%')
                OR cm.chatRoom.caretaker.accountData.surname LIKE CONCAT('%', :chatterData, '%')
            )
        ORDER BY cm.createdAt DESC
    """)
    Page<ChatMessage> findLastMessagesOfChatRoomsOfCaretaker(String caretakerUsername, String chatterData, Pageable pageable);
```

Chyba nic lepszego nie wymyślę xd

Reply:
Tu bym musiał ten ChatMessage pierwszy wyciągnąć żeby race condition nie było jak ktoś wyśle wiadomość podczas GETa

Reply:
Plus tego query że specyfikacje wywale całą

Reply:
A nie możesz zrobić select chatmessage where chatMessage.chatId = chatId, wtedy możesz listę chatId przekazać jaki parametr

Reply:
I wtedy to jest wywoływane bezpośrednio przed zwróceniem odpowiedzi więc wiadomości powinny być najaktualniejsze

Reply:
W sensie pobrać te czaty użytkownika według specification i mając czaty wywołać query dla każdego czaty które przyjmie chatId i na jego podstawie zwroci ostatnia wiadomość z czatu

Reply:
A w którym momencie posortuje czat po ostatniej wiadomości?

Reply:
W tym query do wyciągnięcia ostatniej wiadomości z czatu

Reply:
A bo my jeszcze chcemy żeby czaty w odpowiedniej kolejności się zwrocily

Reply:
Dobra, to rzeczywiście się tak nie da. To może być jak jest

Reply:
Czyli to query czy specyfikacja?

Reply:
Specyfikacja